### PR TITLE
Handle dynamic operations in CFG

### PR DESCRIPTION
**Pull request by mavasani**
Specifically: [IDynamicInvocationOperation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IDynamicInvocationOperation.cs,19), [IDynamicIndexerAccessOperation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IDynamicIndexerAccessOperation.cs,18) and [IDynamicMemberReferenceOperation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IDynamicMemberReferenceOperation.cs,19)

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >return new DynamicInvocationExpression(rewrittenOperation, rewrittenArguments, ((HasDynamicArgumentsExpression)operation).ArgumentNames, ((HasDynamicArgumentsExpression)operation).ArgumentRefKinds, semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit); [](start = 12, length = 300)

Please break this long line into multiple lines that can reasonably fit the screen

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >return new DynamicIndexerAccessExpression(rewrittenOperation, rewrittenArguments, ((HasDynamicArgumentsExpression)operation).ArgumentNames, ((HasDynamicArgumentsExpression)operation).ArgumentRefKinds, semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit); [](start = 12, length = 303)

Please break this long line into multiple lines that can reasonably fit the screen

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >return new DynamicMemberReferenceExpression(Visit(operation.Instance), operation.MemberName, operation.TypeArguments, operation.ContainingType, semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit); [](start = 12, length = 246)

Please break this long line into multiple lines that can reasonably fit the screen

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >VisitDynamicIndexerAccess [](start = 35, length = 25)

It looks like beginning of this function is the same as in VisitDynamicInvocation, consider extracting into a helper method and reusing.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs]: >p = (c1 ?? c2)[d]; [](start = 8, length = 18)

It would be good to have a test covering scenario without any control flow in this construct. It will help us to confirm that there is no unnecessary stack spilling is happening.

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs]: Sure.

---
In reply to: [186830815](https://github.com/dotnet/roslyn/pull/26690#discussion_r186830815) [](ancestors = 186830815)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs]: >d?.Prop1 [](start = 16, length = 8)

It doesn't look like any of the tests are actually testing interesting scenarios with IDynamicMemberReferenceOperation. All they test conditional access and it is not inside IDynamicMemberReferenceOperation, but rather the IDynamicMemberReferenceOperation is inside conditional access. Please replace them with tests that are targeting control flow inside IDynamicMemberReferenceOperation. For example,  ```(d1 ?? d2).Prop1```. We also should have a simple test that doesn't have any control flow.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs]: I'd say we need only two tests:
- no control flow
- control flow in the receiver

---
In reply to: [186833448](https://github.com/dotnet/roslyn/pull/26690#discussion_r186833448) [](ancestors = 186833448)

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs]: Sure, will update.

---
In reply to: [186834523](https://github.com/dotnet/roslyn/pull/26690#discussion_r186834523) [](ancestors = 186834523,186833448)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.vb]: >DynamicMemberReference_ControlFlowInSimpleMemberAccess [](start = 19, length = 54)

Same comments as for C#

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs]: >(c1 ?? c2).M2(d); [](start = 8, length = 17)

Please make sure we have a test without any control flow.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs]: We also need tests covering non-empty TypeArguments and Containing type, could be folded into no control flow test
Also, null receiver, if possible

---
In reply to: [186834783](https://github.com/dotnet/roslyn/pull/26690#discussion_r186834783) [](ancestors = 186834783,186834523,186833448)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs]: >DynamicIndexerAccess_ControlFlowInReceiver [](start = 20, length = 42)

Can we cover null receiver (is it possible) for IDynamicIndexerAccessOperation?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs]: >IDynamicMemberReferenceOperation [](start = 12, length = 32)

Can there be anything else, but an IDynamicMemberReferenceOperation as the IDynamicInvocationOperation.Operation? It would be good to cover scenario like that too.
We should also consider preserving IDynamicMemberReferenceOperation as IDynamicInvocationOperation.Operation instead of spilling it. In that position it is not considered to be a value, please confirm with examining order of evaluation in IL.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs]: > Is it possible to have a dynamic invocation without arguments? Don't we need at least one dynamic argument?

When receiver is dynamic?

---
In reply to: [186839936](https://github.com/dotnet/roslyn/pull/26690#discussion_r186839936) [](ancestors = 186839936,186839282)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb]: >If(d1, d2).M2(c) [](start = 8, length = 16)

It doesn't look like this scenario is testing IDynamicInvocationOperation, consider adding ```call``` at the beginning of the line

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs]: If we change implementation to handle IDynamicMemberReferenceOperation specially, we'll also need tests with type arguments and containing type.

---
In reply to: [186840169](https://github.com/dotnet/roslyn/pull/26690#discussion_r186840169) [](ancestors = 186840169)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb]: >DynamicInvocation_ControlFlowInReceiver [](start = 19, length = 39)

Are we covering scenarios of an invocation of off an object directly and no receiver case?

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb]: >p = [](start = 8, length = 3)

I think you can simplify this scenario by replacing "p =" with "call"

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs]: Added `DynamicIndexerAccess_NoControlFlow`

---
In reply to: [186831047](https://github.com/dotnet/roslyn/pull/26690#discussion_r186831047) [](ancestors = 186831047,186830815)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs]: Added `DynamicIndexerAccess_NullReceiver`. It generates compiler errors but contains an invalid `IDynamicIndexerAccessOperation`.

---
In reply to: [186838334](https://github.com/dotnet/roslyn/pull/26690#discussion_r186838334) [](ancestors = 186838334)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs]: Added `DynamicInvocation_NoControlFlow`

---
In reply to: [186835303](https://github.com/dotnet/roslyn/pull/26690#discussion_r186835303) [](ancestors = 186835303)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs]: Added `DynamicInvocation_ControlFlowNullReceiver` and `DynamicInvocation_ControlFlowInReceiverEmptyArgList`.

---
In reply to: [186840596](https://github.com/dotnet/roslyn/pull/26690#discussion_r186840596) [](ancestors = 186840596,186839936,186839282)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb
Comment (mavasani) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb]: Added `DynamicInvocation_ControlFlowInvocationOffObject` and `DynamicInvocation_ControlFlowNullReceiver`.

---
In reply to: [186843373](https://github.com/dotnet/roslyn/pull/26690#discussion_r186843373) [](ancestors = 186843373)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb
Comment (mavasani) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.vb]: I get `Error BC30454 Expression is not a method.` after doing this replacement. Retaining the current test, please let me know if there is a different way to reduce baseline.

---
In reply to: [186845083](https://github.com/dotnet/roslyn/pull/26690#discussion_r186845083) [](ancestors = 186845083)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs]: Added `DynamicInvocation_ControlFlowWithDynamicTypedMemberReceiver`.

---
In reply to: [186840169](https://github.com/dotnet/roslyn/pull/26690#discussion_r186840169) [](ancestors = 186840169)

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs]: Verified that we don't need to spill IDynamicMemberReferenceOperation. Updated implementation and test baselines. Also added `DynamicIndexerAccessExpression_ControlFlowWithDynamicReceiver` which verified IDynamicMemberReferenceOperation is not spilled for indexer access.

---
In reply to: [186897472](https://github.com/dotnet/roslyn/pull/26690#discussion_r186897472) [](ancestors = 186897472,186840169)

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicInvocationExpression.cs]: Folded non-zero type arguments and non-null containing type verification in test `DynamicInvocation_ControlFlowNullReceiver`

---
In reply to: [186842102](https://github.com/dotnet/roslyn/pull/26690#discussion_r186842102) [](ancestors = 186842102,186840169)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs]: Updated to have only two tests. Null receiver is covered in test `DynamicInvocation_ControlFlowNullReceiver`

---
In reply to: [186837885](https://github.com/dotnet/roslyn/pull/26690#discussion_r186837885) [](ancestors = 186837885,186834783,186834523,186833448)

**Comment by mavasani** [comment]
@AlekseyTs Addressed feedback.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >VisitDynamicOperationWithArguments [](start = 93, length = 34)

It doesn't look like we can use this helper for IDynamicIndexerAccessOperation, it looks like ```Operation``` is considered a value for the purpose of indexing and it should be spilled in case of a control flow in the arguments

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs]: >} [](start = 8, length = 1)

Given the current implementation, it looks like the cases of non-empty type arguments and containing type are not covered. Please make sure we have a test case that covers VisitDynamicMemberReference for these cases.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicIndexerAccessExpression.cs]: >IDynamicMemberReferenceOperation (Member Name: ""M"", Containing Type: null) (OperationKind.DynamicMemberReference, Type: dynamic) (Syntax: 'd.M') [](start = 20, length = 146)

From the IL, it looks like there are two dynamic invocations here and the argument is evaluated between them. It feels like we should spill the IDynamicMemberReferenceOperation node in this case.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Fixed, and verified with test baseline change for `DynamicIndexerAccess_ControlFlowWithDynamicReceiver`

---
In reply to: [187246124](https://github.com/dotnet/roslyn/pull/26690#discussion_r187246124) [](ancestors = 187246124)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDynamicMemberReferenceExpression.cs]: Added test `DynamicMemberReference_ControlFlowInReceiver_TypeArguments` for non-zero type arguments. I am unable to come up with any test where there is non-null containing type, but the parent is not an `IDynamicInvocation`. Can you provide a specific example?

---
In reply to: [187247914](https://github.com/dotnet/roslyn/pull/26690#discussion_r187247914) [](ancestors = 187247914)

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3)

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_spanish_unit32_prtest  please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please

**Comment by mavasani** [comment]
@dotnet retest windows_debug_vs-integration_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_unit32_prtest  please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest please

**Comment by mavasani** [comment]
@333fred @dotnet/roslyn-compiler need second review.

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest  please

**Review by 333fred — COMMENTED** [review]
LGTM (iteration 4)

**Review by 333fred — APPROVED** [review]

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest  please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest  please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_debug_vs-integration_prtest  please

**Pull request merged**
